python/whylogs/core/schema.py

```diff
@@ -161,6 +173,30 @@ def _resolve_pdf(self, df: pd.DataFrame, force_resolve: bool = False) -> bool:
 
         return dirty
 
+    def _augment(
+        self, pandas: Optional[pd.DataFrame] = None, row: Optional[Mapping[str, Any]] = None
+    ) -> Tuple[Optional[pd.DataFrame], Optional[Mapping[str, Any]]]:
+        new_columns = dict(row) if row else None
+        if row is not None:
+            for column, value in row.items():
+                for augmenter in self.augmenters:
+                    col_name, col_type = augmenter.column_name, augmenter.column_type
+                    why_type = self.type_mapper(type(value))
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
+                        for new_col, callback in augmenter.callbacks.items():
+                            new_columns[new_col] = callback(value)  # type: ignore
+
+        if pandas is not None:
+            for column in pandas.keys():
+                for augmenter in self.augmenters:
+                    col_name, col_type = augmenter.column_name, augmenter.column_type
+                    why_type = pandas.dtypes[column]
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
+                        for new_col, callback in augmenter.callbacks.items():
+                            pandas[new_col] = pandas[column].map(callback)
```

Comment:
Let not add the derived columns to the input pandas dataframe which modifies the input. Can we keep these in a separate dataframe?